### PR TITLE
Add a function for downloading static assets for an HTML page

### PR DESCRIPTION
**Pull request by divad12**
Tested with the Blockly, 3asafeer, and Thoughtful Learning sushi chefs. Corresponding PRs to use this function for each of the above will follow.

### CODE REVIEW

File: ricecooker/utils/downloader.py [hunk 1]
```diff
@@ -41,3 +54,180 @@ def read(path, loadjs=False, session=None, driver=None):
...
+        try:
+            response = sess.get(url, headers=headers, timeout=timeout, *args, **kwargs)
+            break
+        except (requests.exceptions.ConnectionError, requests.exceptions.ReadTimeout) as e:
+            retry_count += 1
+            print("Error with connection ('{msg}'); about to perform retry {count} of {trymax}."
+                  .format(msg=str(e), count=retry_count, trymax=max_retries))
+            time.sleep(retry_count * 1)
+            if retry_count >= max_retries:
+                return Dummy404ResponseObject(url=url)
```
Comment (ivanistheone) [inline ricecooker/utils/downloader.py]: The `Dummy404ResponseObject` is missing.

It is defined here:  https://github.com/learningequality/sushi-chef-ck12/blob/master/chef.py#L77
but may be better to refactor to return `None` (or some other way to signal error).

File: ricecooker/utils/downloader.py [hunk 2]
```diff
@@ -41,3 +54,180 @@ def read(path, loadjs=False, session=None, driver=None):
...
+
+
+# TODO(davidhu): Use MD5 hash of URL (ideally file) instead.
+def _derive_filename(url):
+    name = os.path.basename(urlparse(url).path).replace('%', '_')
+    return ("%s.%s" % (uuid.uuid4().hex, name)).lower()
+
+
+def make_fully_qualified_url(url, base):
+    # TODO(davidhu): This is some hacky special-casing for 3asafeer ...
```
Comment (ivanistheone) [inline ricecooker/utils/downloader.py]: I believe some of these can be handled by `urllib.parse.urljoin`, e.g., 
```
from urllib.parse import urljoin

base = 'http://somesite.com/subpath/file.html'

print(urljoin(base, '../image'))
http://somesite.com/image

print(urljoin(base, '../scripts'))
http://somesite.com/scripts

print(urljoin(base, '//some.full.path/just/no/scheme.html'))
http://some.full.path/just/no/scheme.html

print(urljoin(base, '/an/abs/path.html'))
http://somesite.com/an/abs/path.html

print(urljoin(base, 'a/rel/path.html'))
http://somesite.com/subpath/a/rel/path.html
```

File: ricecooker/utils/downloader.py [hunk 3]
```diff
@@ -1,8 +1,15 @@
...
 import requests
 import time
+from urllib.parse import urlparse, parse_qs
+import uuid
+
+from bs4 import BeautifulSoup
 from selenium import webdriver
 from requests_file import FileAdapter
 from ricecooker.config import PHANTOMJS_PATH
+from ricecooker.utils.html import download_file, WebDriver
```
Comment (ivanistheone) [inline ricecooker/utils/downloader.py]: `WebDriver` is not used

Comment (divad12) [reply ricecooker/utils/downloader.py]: Removed

File: ricecooker/utils/downloader.py (hunk 1 above)
Comment (divad12) [reply ricecooker/utils/downloader.py]: Changed to `raise e`

File: ricecooker/utils/downloader.py (hunk 2 above)
Comment (divad12) [reply ricecooker/utils/downloader.py]: This is _awesome_, thanks for sharing!!

**Comment by divad12** [comment]
Thanks for the review @ivanistheone !

> Also, not sure if we can combine the new function with the read method from the other PR? Or maybe keep thing separated.

I think it'd be best to keep things separated, to give users the flexibility of getting a page's HTML without also grabbing all of the static assets, e.g. if we're scraping a video page and want to fetch metadata from the HTML source. Thanks for suggesting though!

**Pull request merged**